File: modules/ai-control/src/main/java/com/acme/opsweave/aicontrol/domain/ModelSpend.java
package com.acme.opsweave.aicontrol.domain;

import com.acme.opsweave.identity.domain.SubjectId;
import com.acme.opsweave.sharedkernel.TenantId;
import java.time.*;
import java.util.*;

/** Local admission/accounting estimates, not a provider invoice or a grant from model output. */
public final class ModelSpend {
    private ModelSpend() {}
    public static final int MAX_INPUT_BYTES = 73_728, MAX_INPUT_TOKENS = 81_920, MAX_OUTPUT_TOKENS = 2_048;
    public static final int MAX_RECORDS = 10_000;
    public static final long MAX_MONEY = 1_000_000_000_000L;
    public record Policy(String provider, String model, String priceVersion, long inputMicrosPerMillion,
            long outputMicrosPerMillion, long maxCallMicros, long dailyMicros) {
        public Policy {
            new InsightSubmission.ModelRef(provider, model);
            if(priceVersion == null || !priceVersion.matches("[a-zA-Z0-9._-]{1,64}")) invalid();
            for(long value : new long[]{inputMicrosPerMillion,outputMicrosPerMillion,maxCallMicros,dailyMicros}) if(value < 0 || value > MAX_MONEY) invalid();
            if(provider.equals("mock-deterministic")) {
                if(!model.equals("mock-current-v1") || !priceVersion.equals("mock-no-charge") || inputMicrosPerMillion != 0 || outputMicrosPerMillion != 0 || maxCallMicros != 0 || dailyMicros != 0) invalid();
            } else if(inputMicrosPerMillion == 0 || outputMicrosPerMillion == 0 || maxCallMicros == 0 || dailyMicros < maxCallMicros) invalid();
        }
        public static Policy mock() { return new Policy("mock-deterministic","mock-current-v1","mock-no-charge",0,0,0,0); }
        public long estimate(int inputTokens, int outputTokens) {
            if(inputTokens < 0 || inputTokens > MAX_INPUT_TOKENS || outputTokens < 0 || outputTokens > MAX_OUTPUT_TOKENS) invalid();
            // Bounded integer arithmetic. Round up once; cached input is conservatively charged at the full input rate.
            return Math.floorDiv(Math.addExact(Math.addExact(Math.multiplyExact((long)inputTokens,inputMicrosPerMillion),
                Math.multiplyExact((long)outputTokens,outputMicrosPerMillion)),999_999),1_000_000);
        }
        public long reservationMicros() { return estimate(MAX_INPUT_TOKENS,MAX_OUTPUT_TOKENS); }
    }
    public record Usage(int inputTokens, int outputTokens, int cachedInputTokens, String source) {
        public Usage {
            if(inputTokens < 0 || inputTokens > MAX_INPUT_TOKENS || outputTokens < 0 || outputTokens > MAX_OUTPUT_TOKENS
                || cachedInputTokens < 0 || cachedInputTokens > inputTokens || !Set.of("provider-reported","mock-no-call").contains(source)) invalid();
            if(source.equals("mock-no-call") ? inputTokens != 0 || outputTokens != 0 || cachedInputTokens != 0 : inputTokens == 0) invalid();
        }
    }
    public record Call(TenantId tenantId, SubjectId subjectId, UUID runId, UUID sessionId, UUID incidentId,
            String inputDigest, int inputBytes, Policy policy, Instant reservedAt, Instant deadlineAt, Usage usage, Instant reportedAt) {
        public Call {
            Objects.requireNonNull(tenantId); Objects.requireNonNull(subjectId); Objects.requireNonNull(runId); Objects.requireNonNull(sessionId); Objects.requireNonNull(incidentId);
            Objects.requireNonNull(policy); Objects.requireNonNull(reservedAt); Objects.requireNonNull(deadlineAt);
            if(inputDigest == null || !inputDigest.matches("sha256:[0-9a-f]{64}") || inputBytes < 1 || inputBytes > MAX_INPUT_BYTES
                || reservedAt.isBefore(Instant.EPOCH) || !deadlineAt.isAfter(reservedAt) || deadlineAt.isAfter(reservedAt.plusSeconds(60))
                || ((usage == null) != (reportedAt == null))) invalid();
            if(usage != null && (reportedAt.isBefore(reservedAt) || !policy.provider().equals(usage.source().equals("mock-no-call") ? "mock-deterministic" : "rig-openai"))) invalid();
        }
        public LocalDate day() { return reservedAt.atOffset(ZoneOffset.UTC).toLocalDate(); }
        public long reservedMicros() { return policy.reservationMicros(); }
        public long chargedMicros() { return usage == null ? reservedMicros() : policy.estimate(usage.inputTokens(),usage.outputTokens()); }
        public boolean countsOn(LocalDate day) { return usage == null || day().equals(day); }
        public String state(Instant now) { return usage != null ? "REPORTED" : now.isBefore(deadlineAt) ? "RESERVED" : "UNCERTAIN"; }
        public Call report(Usage value, Instant now) {
            Objects.requireNonNull(value);
            if(usage != null) { if(!usage.equals(value)) throw new ToolFailure(ToolFailure.Code.INPUT_CHANGED); return this; }
            return new Call(tenantId,subjectId,runId,sessionId,incidentId,inputDigest,inputBytes,policy,reservedAt,deadlineAt,value,now);
        }
    }
    public static void admit(Call call, long accountedMicros, long retainedRecords) {
        if(call.usage()!=null || accountedMicros < 0 || retainedRecords < 0) invalid();
        if(retainedRecords >= MAX_RECORDS || call.reservedMicros() > call.policy().maxCallMicros()
            || accountedMicros > call.policy().dailyMicros() || call.reservedMicros() > call.policy().dailyMicros()-accountedMicros)
            throw new ToolFailure(ToolFailure.Code.BUDGET_EXHAUSTED);
    }
    private static void invalid() { throw new ToolFailure(ToolFailure.Code.INVALID_REQUEST); }
}

File: apps/web-console/src/api/model-spend.ts
import { platformClient } from './http.ts'
import { isUuid, InsightRequestError, type Insight } from './insights.ts'

// Explicit handwritten boundary for contracts/schemas/v1/model-spend-*. No floating-point billing arithmetic.
type Usage = { inputTokens: number; outputTokens: number; cachedInputTokens: number; source: 'mock-no-call' | 'provider-reported' }
export type ModelSpend = { storage: 'memory' | 'postgres'; record: {
  schemaVersion: '1.0'; tenantId: string; subjectId: string; runId: string; sessionId: string; incidentId: string;
  inputDigest: string; inputBytes: number; reservedAt: string; deadlineAt: string; reportedAt: string | null;
  state: 'RESERVED' | 'UNCERTAIN' | 'REPORTED'; usage: Usage | null; reservedMicros: number; estimatedMicros: number | null; accountedMicros: number;
  currency: 'USD'; maxInputTokens: 81920; maxOutputTokens: 2048; maxInputBytes: 73728; accounting: 'configured-price-estimate';
  policy: { provider: 'mock-deterministic' | 'rig-openai'; model: string; priceVersion: string; inputMicrosPerMillion: number; outputMicrosPerMillion: number; maxCallMicros: number; dailyMicros: number }
} }
function check(value: unknown): asserts value { if (!value) throw new Error('用量响应结构、费用或范围不正确') }
function exact(v: unknown, keys: string[]): asserts v is Record<string, unknown> { check(v !== null && typeof v === 'object' && !Array.isArray(v) && Object.keys(v).length === keys.length && keys.every(k => k in v)) }
const integer = (v: unknown, max: number, min = 0): v is number => Number.isSafeInteger(v) && Number(v) >= min && Number(v) <= max
const uuid = (v: unknown): v is string => isUuid(v) && v.length === 36
const str = (v: unknown, max = 128): v is string => typeof v === 'string' && [...v].length > 0 && [...v].length <= max
const date = (v: unknown): v is string => typeof v === 'string' && /^\d{4}-\d\d-\d\dT\d\d:\d\d:\d\d(?:\.\d{1,9})?Z(?![\s\S])/.test(v) && Number.isFinite(Date.parse(v))
export function parseModelSpend(v: unknown, runId: string, insight?: Insight): ModelSpend {
  exact(v, ['storage', 'record']); check(['memory', 'postgres'].includes(String(v.storage)))
  const r = v.record; exact(r, ['schemaVersion', 'tenantId', 'subjectId', 'runId', 'sessionId', 'incidentId', 'inputDigest', 'inputBytes', 'policy', 'reservedAt', 'deadlineAt', 'reportedAt', 'usage', 'state', 'currency', 'reservedMicros', 'estimatedMicros', 'accountedMicros', 'maxInputTokens', 'maxOutputTokens', 'maxInputBytes', 'accounting'])
  check(r.schemaVersion === '1.0' && r.runId === runId && uuid(r.runId) && uuid(r.sessionId) && uuid(r.incidentId) && str(r.tenantId) && str(r.subjectId)
    && typeof r.inputDigest === 'string' && /^sha256:[0-9a-f]{64}(?![\s\S])/.test(r.inputDigest) && integer(r.inputBytes, 73728, 1) && r.maxInputBytes === 73728
    && r.maxInputTokens === 81920 && r.maxOutputTokens === 2048 && r.currency === 'USD' && r.accounting === 'configured-price-estimate')
  check(date(r.reservedAt) && date(r.deadlineAt) && Date.parse(r.reservedAt) >= 0 && Date.parse(r.reservedAt) <= Date.now() && Date.parse(r.reservedAt) < Date.parse(r.deadlineAt) && Date.parse(r.deadlineAt) - Date.parse(r.reservedAt) <= 60000)
  const p = r.policy; exact(p, ['provider', 'model', 'priceVersion', 'inputMicrosPerMillion', 'outputMicrosPerMillion', 'maxCallMicros', 'dailyMicros'])
  check(['mock-deterministic', 'rig-openai'].includes(String(p.provider)) && str(p.model) && typeof p.priceVersion === 'string' && /^[a-zA-Z0-9._-]{1,64}(?![\s\S])/.test(p.priceVersion))
  check(integer(p.inputMicrosPerMillion, 1e12) && integer(p.outputMicrosPerMillion, 1e12) && integer(p.maxCallMicros, 1e12) && integer(p.dailyMicros, 1e12) && p.dailyMicros >= p.maxCallMicros)
  if (p.provider === 'mock-deterministic') check(p.model === 'mock-current-v1' && p.priceVersion === 'mock-no-charge' && p.inputMicrosPerMillion === 0 && p.outputMicrosPerMillion === 0 && p.maxCallMicros === 0 && p.dailyMicros === 0)
  else check(v.storage === 'postgres' && p.inputMicrosPerMillion > 0 && p.outputMicrosPerMillion > 0 && p.maxCallMicros > 0)
  const estimate = (input: number, output: number) => Number((BigInt(input) * BigInt(p.inputMicrosPerMillion as number) + BigInt(output) * BigInt(p.outputMicrosPerMillion as number) + 999999n) / 1000000n)
  check(integer(r.reservedMicros, 1e12) && r.reservedMicros === estimate(81920, 2048) && r.reservedMicros <= p.maxCallMicros && integer(r.accountedMicros, 1e12))
  if (r.state === 'REPORTED') {
    check(date(r.reportedAt) && Date.parse(r.reportedAt) >= Date.parse(r.reservedAt) && Date.parse(r.reportedAt) <= Date.now())
    const u = r.usage; exact(u, ['inputTokens', 'outputTokens', 'cachedInputTokens', 'source'])
    check(integer(u.inputTokens, 81920) && integer(u.outputTokens, 2048) && integer(u.cachedInputTokens, u.inputTokens))
    if (p.provider === 'mock-deterministic') check(u.source === 'mock-no-call' && u.inputTokens === 0 && u.outputTokens === 0 && u.cachedInputTokens === 0)
    else check(u.source === 'provider-reported' && u.inputTokens > 0)
    check(r.estimatedMicros === estimate(u.inputTokens, u.outputTokens) && r.accountedMicros === r.estimatedMicros)
  } else check(['RESERVED', 'UNCERTAIN'].includes(String(r.state)) && r.usage === null && r.reportedAt === null && r.estimatedMicros === null && r.accountedMicros === r.reservedMicros && (r.state !== 'UNCERTAIN' || Date.parse(r.deadlineAt) <= Date.now()))
  if (insight) check(r.runId === insight.id && r.sessionId === insight.sessionId && r.tenantId === insight.tenantId && r.subjectId === insight.subjectId && r.incidentId === insight.incidentId && p.provider === insight.model.provider && p.model === insight.model.name && r.state === 'REPORTED')
  return v as unknown as ModelSpend
}
export async function getModelSpend(id: string, signal: AbortSignal, insight?: Insight) {
  check(uuid(id)); const value = await platformClient.request(`/api/v1/ai/model-calls/${id}`, { signal, responseBytes: 16384, timeoutMs: 18000,
    error: status => status === 404 ? new Error('尚未找到用量记录；无法据此确认费用为零，也不会自动重试模型。') : new InsightRequestError(status) })
  return parseModelSpend(value, id, insight)
}
export const dollars = (micros: number) => `USD ${(micros / 1000000).toFixed(6)}`
